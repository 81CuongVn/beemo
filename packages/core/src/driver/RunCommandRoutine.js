/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Routine } from 'boost';
import chalk from 'chalk';
import glob from 'glob';
import isGlob from 'is-glob';
import parseArgs from 'yargs-parser';

import type { DriverContext, Execution } from '../types';

type Args = string[];

const OPTION_PATTERN: RegExp = /-?-[-a-z0-9]+(,|\s)/gi;

export default class RunCommandRoutine extends Routine<Object, DriverContext> {
  execute(): Promise<*> {
    const { metadata } = this.context.primaryDriver;

    this.task('Gathering arguments', this.gatherArgs);

    this.task('Expanding glob patterns', this.expandGlobPatterns);

    this.task('Filtering options', this.filterUnknownOptions).skip(!metadata.filterOptions);

    this.task('Including config option', this.includeConfigOption).skip(!metadata.useConfigOption);

    this.task('Running command', this.runCommandWithArgs);

    return this.serializeTasks();
  }

  /**
   * Expand arguments that look like globs.
   */
  expandGlobPatterns(args: Args): Promise<Args> {
    const nextArgs = [];

    this.tool.debug('Expanding glob patterns');

    args.forEach(arg => {
      if (isGlob(arg)) {
        const paths = glob.sync(arg, {
          cwd: this.context.root,
          debug: this.tool.config.debug,
          strict: true,
        });

        this.tool.debug(`  ${arg} ${chalk.gray('->')} ${paths.join(', ')}`);

        nextArgs.push(...paths);
      } else {
        nextArgs.push(arg);
      }
    });

    return Promise.resolve(nextArgs);
  }

  /**
   * Extract native supported options and flags from driver help output.
   */
  extractNativeOptions(): Promise<{ [option: string]: true }> {
    const driver = this.context.primaryDriver;
    const { env } = driver.options;
    const options = driver.getSupportedOptions();

    if (options.length > 0) {
      this.tool.debug('Using supported options from driver');

      const nativeOptions = {};

      options.forEach(option => {
        nativeOptions[option] = true;
      });

      return Promise.resolve(nativeOptions);
    }

    this.tool.debug('Extracting native options from help output');

    return this.executeCommand(driver.metadata.bin, [driver.metadata.helpOption], { env }).then(
      ({ stdout }) => {
        const nativeOptions = {};

        stdout.match(OPTION_PATTERN).forEach(option => {
          // Trim trailing comma or space
          nativeOptions[option.slice(0, -1)] = true;
        });

        return nativeOptions;
      },
    );
  }

  /**
   * Filter unknown and or unsupported CLI options from the arguments passed to the CLI.
   * Utilize the driver's help option/command to determine accurate options.
   */
  filterUnknownOptions(args: Args): Promise<Args> {
    this.tool.debug('Filtering unknown command line options');

    return this.extractNativeOptions().then(nativeOptions => {
      const filteredArgs = [];
      const unknownArgs = [];
      let skipNext = false;

      args.forEach((arg, i) => {
        if (skipNext) {
          skipNext = false;

          return;
        }

        if (arg.startsWith('-')) {
          let option = arg;
          const nextArg = args[i + 1];

          // --opt=123
          if (option.includes('=')) {
            [option] = option.split('=');

            if (!nativeOptions[option]) {
              unknownArgs.push(arg);

              return;
            }

            // --opt 123
          } else if (!nativeOptions[option]) {
            unknownArgs.push(arg);

            if (nextArg && !nextArg.startsWith('-')) {
              skipNext = true;
              unknownArgs.push(nextArg);
            }

            return;
          }
        }

        filteredArgs.push(arg);
      });

      if (unknownArgs.length > 0) {
        this.tool.debug(`Filtered args: ${unknownArgs.join(', ')}`);
      }

      return filteredArgs;
    });
  }

  /**
   * Gather arguments from all sources to pass to the driver.
   */
  gatherArgs(): Promise<Args> {
    const driverArgs = this.context.primaryDriver.getArgs();
    const commandArgs = this.context.args;
    const args = [
      // Passed by the driver
      ...driverArgs,
      // Passed on the command line
      ...commandArgs,
      // Parallel args passed on the command line
      // ...(this.config.parallelArgs || []),
    ];

    this.tool.debug('Gathering arguments to pass to driver');

    this.tool.invariant(
      driverArgs.length > 0,
      '  From driver "args" option',
      driverArgs.join(' '),
      'No arguments',
    );

    this.tool.invariant(
      commandArgs.length > 0,
      '  From the command line',
      commandArgs.join(' '),
      'No arguments',
    );

    // Since we combine multiple args, we need to rebuild this.
    // And we also need to set this before we filter them.
    this.context.yargs = parseArgs(args);

    return Promise.resolve(args);
  }

  /**
   * Include --config option if driver requires it (instead of auto-lookup resolution).
   */
  includeConfigOption(prevArgs: Args): Promise<Args> {
    const { configPaths, primaryDriver } = this.context;
    const configPath = configPaths.find(path => path.endsWith(primaryDriver.metadata.configName));
    const args = [...prevArgs];

    if (configPath) {
      args.push(primaryDriver.metadata.configOption, configPath);
    }

    this.tool.debug('Including config option to args');

    return Promise.resolve(args);
  }

  /**
   * Execute the driver's command with the filtered arguments and handle the
   * success and failures with the driver itself.
   */
  runCommandWithArgs(args: Args): Promise<Execution> {
    const driver = this.context.primaryDriver;

    this.tool.debug(
      `Executing command ${chalk.magenta(driver.metadata.bin)} with args "${args.join(' ')}"`,
    );

    this.tool.emit('before-execute', [driver, args, this.context]);

    return this.executeCommand(driver.metadata.bin, args, { env: driver.options.env })
      .then(response => {
        driver.handleSuccess(response);

        this.tool.emit('after-execute', [driver, response]);

        return response;
      })
      .catch(error => {
        driver.handleFailure(error);

        this.tool.emit('failed-execute', [driver, error]);

        throw error;
      });
  }
}
