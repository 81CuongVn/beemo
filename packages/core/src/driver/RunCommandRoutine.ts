/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Routine, Task } from '@boost/core';
import chalk from 'chalk';
import glob from 'fast-glob';
import path from 'path';
import fs from 'fs-extra';
import isGlob from 'is-glob';
import merge from 'lodash/merge';
import optimal, { array, bool, string } from 'optimal';
import parseArgs from 'yargs-parser';
import DriverContext from '../contexts/DriverContext';
import { STRATEGY_COPY } from '../constants';
import { Argv, Execution, BeemoTool } from '../types';

const OPTION_PATTERN: RegExp = /-?-[a-z0-9-]+(,|\s)/giu;

export type OptionMap = { [option: string]: true };

export interface RunCommandOptions {
  additionalArgv: Argv;
  forceConfigOption: boolean;
  workspaceRoot: string;
}

export default class RunCommandRoutine extends Routine<
  DriverContext,
  BeemoTool,
  RunCommandOptions
> {
  bootstrap() {
    this.options = optimal(
      this.options,
      {
        additionalArgv: array(string()),
        forceConfigOption: bool(),
        workspaceRoot: string().empty(),
      },
      {
        name: 'RunCommandRoutine',
      },
    );
  }

  async execute(context: DriverContext): Promise<Execution> {
    const { tool } = this;
    const { forceConfigOption, workspaceRoot } = this.options;
    const { metadata } = context.primaryDriver;

    this.task(tool.msg('app:driverRunGatherArgs'), this.gatherArgs);

    this.task(tool.msg('app:driverRunExpandGlob'), this.expandGlobPatterns);

    this.task(tool.msg('app:driverRunFilterOptions'), this.filterUnknownOptions).skip(
      !metadata.filterOptions,
    );

    if (workspaceRoot && metadata.workspaceStrategy === STRATEGY_COPY) {
      this.task(tool.msg('app:driverRunCopyWorkspaceConfig'), this.copyConfigToWorkspace);
    } else {
      this.task(tool.msg('app:driverRunIncludeConfigOption'), this.includeConfigOption).skip(
        !metadata.useConfigOption && !forceConfigOption,
      );
    }

    this.task(tool.msg('app:driverRunCommand'), this.runCommandWithArgs);

    return this.serializeTasks();
  }

  /**
   * When workspaces are enabled, some drivers require the config to be within each workspace,
   * instead of being referenced from the root, so we need to copy it.
   */
  async copyConfigToWorkspace(context: DriverContext, argv: Argv): Promise<Argv> {
    const { workspaceRoot } = this.options;

    this.debug('Copying config files to workspace');

    context.configPaths.forEach(config => {
      fs.copyFileSync(config.path, path.join(workspaceRoot, path.basename(config.path)));
    });

    return argv;
  }

  /**
   * Expand arguments that look like globs.
   */
  async expandGlobPatterns(context: DriverContext, argv: Argv): Promise<Argv> {
    const nextArgv: Argv = [];

    this.debug('Expanding glob patterns');

    argv.forEach(arg => {
      if (arg.charAt(0) !== '-' && isGlob(arg)) {
        const paths = glob
          .sync(arg, {
            cwd: context.root,
            onlyDirectories: false,
            onlyFiles: false,
          })
          .map(filePath => String(filePath));

        this.debug(
          '  %s %s %s',
          arg,
          chalk.gray('->'),
          paths.length > 0 ? paths.join(', ') : chalk.gray(this.tool.msg('app:noMatch')),
        );

        nextArgv.push(...paths);
      } else {
        nextArgv.push(arg);
      }
    });

    return nextArgv;
  }

  /**
   * Extract native supported options and flags from driver help output.
   */
  async extractNativeOptions(): Promise<OptionMap> {
    const driver = this.context.primaryDriver;
    const { env } = driver.options;
    const options = driver.getSupportedOptions();

    if (options.length > 0) {
      this.debug('Using supported options from driver');

      const nativeOptions: OptionMap = {};

      options.forEach(option => {
        nativeOptions[option] = true;
      });

      return Promise.resolve(nativeOptions);
    }

    this.debug('Extracting native options from help output');

    const { stdout } = await this.executeCommand(
      driver.metadata.bin,
      driver.metadata.helpOption.split(' '),
      {
        env,
      },
    );

    const nativeOptions: OptionMap = {};
    const matches = stdout.match(OPTION_PATTERN) || [];

    matches.forEach(option => {
      // Trim trailing comma or space
      nativeOptions[option.slice(0, -1)] = true;
    });

    return nativeOptions;
  }

  /**
   * Filter unknown and or unsupported CLI options from the arguments passed to the CLI.
   * Utilize the driver's help option/command to determine accurate options.
   */
  async filterUnknownOptions(context: DriverContext, argv: Argv): Promise<Argv> {
    this.debug('Filtering unknown command line options');

    const nativeOptions = await this.extractNativeOptions();
    const filteredArgv: Argv = [];
    const unknownArgv: Argv = [];
    let skipNext = false;

    argv.forEach((arg, i) => {
      if (skipNext) {
        skipNext = false;

        return;
      }

      if (arg.startsWith('-')) {
        let option = arg;
        const nextArg = argv[i + 1];

        // --opt=123
        if (option.includes('=')) {
          [option] = option.split('=');

          if (!nativeOptions[option]) {
            unknownArgv.push(arg);

            return;
          }

          // --opt 123
        } else if (!nativeOptions[option]) {
          unknownArgv.push(arg);

          if (nextArg && !nextArg.startsWith('-')) {
            skipNext = true;
            unknownArgv.push(nextArg);
          }

          return;
        }
      }

      filteredArgv.push(arg);
    });

    if (unknownArgv.length > 0) {
      this.debug('Filtered args: %s', chalk.gray(unknownArgv.join(', ')));
    }

    return filteredArgv;
  }

  /**
   * Gather arguments from all sources to pass to the driver.
   */
  async gatherArgs(context: DriverContext): Promise<Argv> {
    this.debug('Gathering arguments to pass to driver');

    const argv = [
      // Passed by the driver
      ...this.getDriverArgs(),
      // Passed on the command line
      ...this.getCommandLineArgs(),
      // Passed with parallel "//" operator
      ...this.getAdditionalArgs(),
    ];

    // Since we combine multiple args, we need to rebuild this.
    // And we need to set this before we filter them.
    // And we need to be sure not to remove existing args.
    context.args = merge({}, parseArgs(argv), context.args);

    return argv;
  }

  /**
   * Run some validation on additional/parallel args.
   */
  getAdditionalArgs(): Argv {
    const argv = this.options.additionalArgv;

    this.debug.invariant(argv.length > 0, 'From parallel operator', argv.join(' '), 'No arguments');

    return argv;
  }

  /**
   * Return args from the command line.
   */
  getCommandLineArgs(): Argv {
    const { argv } = this.context;

    this.debug.invariant(argv.length > 0, 'From the command line', argv.join(' '), 'No arguments');

    return argv;
  }

  /**
   * Return args from the primary driver.
   */
  getDriverArgs(): Argv {
    const argv = this.context.primaryDriver.getArgs();

    this.debug.invariant(
      argv.length > 0,
      'From driver "args" option',
      argv.join(' '),
      'No arguments',
    );

    return argv;
  }

  /**
   * Include --config option if driver requires it (instead of auto-lookup resolution).
   */
  async includeConfigOption(context: DriverContext, prevArgv: Argv): Promise<Argv> {
    const { primaryDriver } = context;
    const configPath = context.findConfigByName(primaryDriver.metadata.configName);
    const argv = [...prevArgv];

    if (configPath && primaryDriver.metadata.configOption) {
      argv.push(primaryDriver.metadata.configOption, configPath.path);
    }

    this.debug('Including config option to args');

    return argv;
  }

  /**
   * Execute the driver's command with the filtered arguments and handle the
   * success and failures with the driver itself.
   */
  async runCommandWithArgs(
    context: DriverContext,
    argv: Argv,
    task: Task<DriverContext>,
  ): Promise<Execution> {
    const driver = context.primaryDriver;
    const cwd = this.options.workspaceRoot || context.root;
    let result = null;

    this.debug(
      'Executing command "%s %s" in %s',
      chalk.magenta(driver.metadata.bin),
      argv.join(' '),
      chalk.cyan(cwd),
    );

    this.tool.emit(`${driver.name}.before-execute`, [context, argv, driver]);

    try {
      result = await this.executeCommand(driver.metadata.bin, argv, {
        cwd,
        env: driver.options.env,
        task,
      });

      driver.handleSuccess(result);

      this.tool.emit(`${driver.name}.after-execute`, [context, result, driver]);
    } catch (error) {
      driver.handleFailure(error);

      this.tool.emit(`${driver.name}.failed-execute`, [context, error, driver]);

      throw error;
    }

    return result;
  }
}
