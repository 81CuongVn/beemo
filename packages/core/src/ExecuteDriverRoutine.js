/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Routine } from 'boost';
import RunCommandRoutine from './driver/RunCommandRoutine';

import type { BeemoConfig, DriverContext } from './types';

export default class ExecuteDriverRoutine extends Routine<BeemoConfig, DriverContext> {
  execute(context: DriverContext): Promise<string[]> {
    const { args, primaryDriver } = context;
    const driverName = primaryDriver.name;
    const binName = primaryDriver.metadata.bin;

    // if (parallel.length > 0) {
    //   const filteredArgs = args.filter(arg => !arg.startsWith('--parallel'));
    //
    //   parallel.forEach(extraArgs => {
    //     const parallelArgs = extraArgs.split(' ');
    //     const combinedArgs = [...filteredArgs, ...parallelArgs];
    //
    //     this.pipe(
    //       new RunCommandRoutine(driverName, `${binName} ${combinedArgs.join(' ')}`, {
    //         parallelArgs,
    //       }),
    //     );
    //   });
    // } else {
    this.pipe(new RunCommandRoutine(driverName, `${binName} ${args.join(' ')}`));
    // }

    return this.parallelizeSubroutines();
  }
}
