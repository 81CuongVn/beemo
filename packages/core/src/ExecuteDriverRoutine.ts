/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import fs from 'fs-extra';
import path from 'path';
import glob from 'glob';
import { Routine, PackageConfig } from '@boost/core';
import DriverContext from './contexts/DriverContext';
import RunCommandRoutine, { RunCommandOptions } from './driver/RunCommandRoutine';
import isPatternMatch from './utils/isPatternMatch';

export default class ExecuteDriverRoutine extends Routine<DriverContext> {
  workspacePackages: PackageConfig[] = [];

  bootstrap() {
    const { args, primaryDriver, workspaces } = this.context;

    if (args.workspaces) {
      if (!workspaces || workspaces.length === 0) {
        throw new Error(
          `Option --workspaces=${args.workspaces} provided but project is not workspaces enabled.`,
        );
      }

      this.workspacePackages = this.loadWorkspacePackages();

      this.getFilteredWorkspaces().forEach(pkg => {
        this.pipeParallelBuilds(pkg.workspaceName, {
          forceConfigOption: true,
          workspaceRoot: pkg.workspacePath,
        });
      });
    } else {
      this.pipeParallelBuilds(primaryDriver.name);
    }
  }

  async execute(context: DriverContext): Promise<string[]> {
    const { other, priority } = this.orderByWorkspacePriorityGraph();

    await this.serializeRoutines(null, priority);

    const response = await this.poolRoutines(
      null,
      {
        concurrency: context.args.concurrency,
      },
      other,
    );

    if (response.errors.length > 0) {
      const messages = response.errors.map(error => error.message);

      throw new Error(`Execution failure.\n${messages.join('\n\n')}`);
    }

    return response.results;
  }

  /**
   * Return a list of workspaces optionally filtered.
   */
  getFilteredWorkspaces(): PackageConfig[] {
    return this.workspacePackages.filter(pkg =>
      // @ts-ignore Contains not typed yet.
      isPatternMatch(pkg.name, this.context.args.workspaces, { contains: true }),
    );
  }

  /**
   * Load package.json from each workspace package.
   */
  loadWorkspacePackages(): PackageConfig[] {
    return glob
      .sync(`${this.context.workspaces}/package.json`, {
        absolute: true,
        cwd: this.context.root,
        debug: this.tool.config.debug,
        strict: true,
      })
      .map(filePath => {
        const pkg = JSON.parse(fs.readFileSync(filePath, 'utf8'));

        pkg.packagePath = filePath;
        pkg.workspacePath = path.dirname(filePath);
        pkg.workspaceName = path.basename(pkg.workspacePath);

        return pkg;
      });
  }

  /**
   * Group routines in order of which they are dependend on.
   */
  orderByWorkspacePriorityGraph(): {
    other: Routine<DriverContext>[];
    priority: Routine<DriverContext>[];
  } {
    if (!this.context.args.priority) {
      return {
        other: this.routines,
        priority: [],
      };
    }

    // Create a mapping of package names within all workspaces
    const packages: { [name: string]: PackageConfig } = {};

    this.workspacePackages.forEach(pkg => {
      packages[pkg.name] = pkg;
    });

    // Determine dependend on packages by resolving the graph and incrementing counts
    const depCounts: { [name: string]: { count: number; package: PackageConfig } } = {};

    this.workspacePackages.forEach(pkg => {
      const deps = {
        ...pkg.dependencies,
        ...pkg.peerDependencies,
      };

      Object.keys(deps).forEach(depName => {
        if (!packages[depName]) {
          return;
        }

        if (depCounts[depName]) {
          depCounts[depName].count += 1;
        } else {
          depCounts[depName] = {
            count: 1,
            package: packages[depName],
          };
        }
      });
    });

    // Order by highest count
    const orderedDeps = Object.values(depCounts)
      .sort((a, b) => b.count - a.count)
      .map(dep => dep.package);

    // Extract dependents in order
    const priority: Routine<DriverContext>[] = [];

    orderedDeps.forEach(pkg => {
      const routine = this.routines.find(route => route.key === pkg.workspaceName);

      if (routine) {
        priority.push(routine);
      }
    });

    // Extract dependers
    const other: Routine<DriverContext>[] = [];

    this.routines.forEach(routine => {
      const dependency = orderedDeps.find(dep => dep.workspaceName === routine.key);

      if (!dependency) {
        other.push(routine);
      }
    });

    return {
      other,
      priority,
    };
  }

  /**
   * When a parallel pipe "|>" is defined, we need to create an additional routine
   * for each instance.
   */
  pipeParallelBuilds(key: string, options: Partial<RunCommandOptions> = {}) {
    const { argv, parallelArgv, primaryDriver } = this.context;
    const command = `${primaryDriver.metadata.bin} ${argv.join(' ')}`;

    if (parallelArgv.length > 0) {
      parallelArgv.forEach(pargv => {
        this.pipe(
          new RunCommandRoutine(key, `${command} ${pargv.join(' ')}`, {
            ...options,
            additionalArgv: pargv,
          }),
        );
      });
    } else {
      this.pipe(new RunCommandRoutine(key, command, options));
    }
  }
}
