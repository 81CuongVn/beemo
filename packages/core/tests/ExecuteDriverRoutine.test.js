import path from 'path';
import { Tool } from 'boost';
import ModuleLoader from 'boost/lib/ModuleLoader';
import ExecuteDriverRoutine from '../src/ExecuteDriverRoutine';
import RunCommandRoutine from '../src/driver/RunCommandRoutine';
import {
  createDriver,
  createDriverContext,
  setupMockTool,
  prependRoot,
} from '../../../tests/helpers';

jest.mock('boost/lib/Tool');

jest.mock('../src/driver/RunCommandRoutine', () => jest.fn());

describe('ExecuteDriverRoutine', () => {
  let routine;
  let driver;

  beforeEach(() => {
    const tool = new Tool();

    driver = createDriver('primary', tool);

    routine = new ExecuteDriverRoutine('driver', 'Executing driver');
    routine.context = createDriverContext(driver);
    routine.tool = setupMockTool(tool);
    routine.debug = jest.fn();

    RunCommandRoutine.mockClear();
  });

  describe('bootstrap()', () => {
    it('adds a routine for the primary driver', () => {
      routine.pipe = jest.fn();
      routine.bootstrap();

      expect(routine.pipe).toHaveBeenCalledWith(
        new RunCommandRoutine('primary', 'primary -a --foo bar baz'),
      );
    });

    it('adds multiple routines when --parallel is used', () => {
      routine.context.args.parallel = ['"--one --two=2"', '" --three -f "'];
      routine.context.argv.push('--parallel="--one --two=2"', '--parallel=" --three -f "');
      routine.pipe = jest.fn();
      routine.bootstrap();

      expect(routine.pipe).toHaveBeenCalledWith(
        new RunCommandRoutine('primary', 'primary -a --foo bar baz --one --two=2', {
          additionalArgv: ['--one', '--two=2'],
        }),
      );
      expect(routine.pipe).toHaveBeenCalledWith(
        new RunCommandRoutine('primary', 'primary -a --foo bar baz --three -f', {
          additionalArgv: ['--three', '-f'],
        }),
      );
    });

    it('adds a routine if parallel is empty', () => {
      routine.context.args.parallel = [];
      routine.pipe = jest.fn();
      routine.bootstrap();

      expect(routine.pipe).toHaveBeenCalledWith(
        new RunCommandRoutine('primary', 'primary -a --foo bar baz'),
      );
    });

    describe('workspaces', () => {
      beforeEach(() => {
        routine.context.args.workspaces = '*';
        routine.context.workspaces = ['packages/*'];
        routine.getWorkspaceFilteredPaths = () => [
          './packages/foo',
          './packages/bar',
          './packages/baz',
        ];
      });

      it('adds a routine for each', () => {
        routine.pipe = jest.fn();
        routine.bootstrap();

        expect(routine.pipe).toHaveBeenCalledTimes(3);
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('foo', 'primary -a --foo bar baz', {
            forceConfigOption: true,
            workspaceRoot: './packages/foo',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('bar', 'primary -a --foo bar baz', {
            forceConfigOption: true,
            workspaceRoot: './packages/bar',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('baz', 'primary -a --foo bar baz', {
            forceConfigOption: true,
            workspaceRoot: './packages/baz',
          }),
        );
      });

      it('adds a routine for each when --parallel is used', () => {
        routine.context.args.parallel = ['"--one --two=2"', '" --three -f "'];
        routine.context.argv.push('--parallel="--one --two=2"', '--parallel=" --three -f "');
        routine.pipe = jest.fn();
        routine.bootstrap();

        expect(routine.pipe).toHaveBeenCalledTimes(6);
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('foo', 'primary -a --foo bar baz --one --two=2', {
            additionalArgv: ['--one', '--two=2'],
            forceConfigOption: true,
            workspaceRoot: './packages/foo',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('foo', 'primary -a --foo bar baz --three -f', {
            additionalArgv: ['--three', '-f'],
            forceConfigOption: true,
            workspaceRoot: './packages/foo',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('bar', 'primary -a --foo bar baz --one --two=2', {
            additionalArgv: ['--one', '--two=2'],
            forceConfigOption: true,
            workspaceRoot: './packages/bar',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('bar', 'primary -a --foo bar baz --three -f', {
            additionalArgv: ['--three', '-f'],
            forceConfigOption: true,
            workspaceRoot: './packages/bar',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('baz', 'primary -a --foo bar baz --one --two=2', {
            additionalArgv: ['--one', '--two=2'],
            forceConfigOption: true,
            workspaceRoot: './packages/baz',
          }),
        );
        expect(routine.pipe).toHaveBeenCalledWith(
          new RunCommandRoutine('baz', 'primary -a --foo bar baz --three -f', {
            additionalArgv: ['--three', '-f'],
            forceConfigOption: true,
            workspaceRoot: './packages/baz',
          }),
        );
      });

      it('errors if workspaces config is not set', () => {
        expect(() => {
          routine.context.workspaces = null;
          routine.bootstrap();
        }).toThrowErrorMatchingSnapshot();
      });

      it('errors if workspaces config is empty', () => {
        expect(() => {
          routine.context.workspaces = [];
          routine.bootstrap();
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('execute()', () => {
    beforeEach(() => {
      // RunCommandRoutine is mocked, so use plain objects
      routine.routines = [
        { key: 'primary' },
        { key: 'foo' },
        { key: 'bar' },
        { key: 'baz' },
        { key: 'qux' },
      ];
    });

    it('pools each routine', async () => {
      routine.poolRoutines = jest.fn(() => Promise.resolve({ errors: [], results: [] }));

      const response = await routine.execute();

      expect(routine.poolRoutines).toHaveBeenCalledWith(null, {}, routine.routines);
    });

    it('throws an error if any failures', async () => {
      routine.poolRoutines = jest.fn(() =>
        Promise.resolve({ errors: [new Error('Failed'), new Error('Oops')], results: [] }),
      );

      try {
        await routine.execute();
      } catch (error) {
        expect(error).toEqual(new Error('Execution failure.\nFailed\n\nOops'));
      }
    });

    it('returns results', async () => {
      routine.poolRoutines = jest.fn(() => Promise.resolve({ errors: [], results: [123] }));

      const response = await routine.execute();

      expect(response).toEqual([123]);
    });

    it('serializes high-priority routines before pooling routines', async () => {
      routine.context.args.priority = 'qux,foo';
      routine.serializeRoutines = jest.fn(() => Promise.resolve());
      routine.poolRoutines = jest.fn(() => Promise.resolve({ errors: [], results: [] }));

      const response = await routine.execute();

      expect(routine.serializeRoutines).toHaveBeenCalledWith(null, [
        { key: 'qux' },
        { key: 'foo' },
      ]);
      expect(routine.poolRoutines).toHaveBeenCalledWith(null, {}, [
        { key: 'primary' },
        { key: 'bar' },
        { key: 'baz' },
      ]);
    });
  });

  describe('getWorkspaceFilteredPaths()', () => {
    it('returns a list of paths', () => {
      routine.context.args.workspaces = '*';
      routine.context.workspaces = ['packages/*'];
      routine.context.root = process.cwd();

      expect(routine.getWorkspaceFilteredPaths()).toEqual(
        expect.arrayContaining([
          path.join(process.cwd(), 'packages/cli'),
          path.join(process.cwd(), 'packages/core'),
          path.join(process.cwd(), 'packages/driver-babel'),
          path.join(process.cwd(), 'packages/driver-typescript'),
        ]),
      );
    });

    it('returns empty if nothing found', () => {
      routine.context.args.workspaces = '*';
      routine.context.workspaces = ['packages/*'];
      routine.context.root = path.join(process.cwd(), '/some/fake/path');

      expect(routine.getWorkspaceFilteredPaths()).toEqual([]);
    });

    it('filters list using --workspaces arg', () => {
      routine.context.args.workspaces = 'driver-*';
      routine.context.workspaces = ['packages/*'];
      routine.context.root = process.cwd();

      expect(routine.getWorkspaceFilteredPaths()).toEqual(
        expect.arrayContaining([
          path.join(process.cwd(), 'packages/driver-babel'),
          path.join(process.cwd(), 'packages/driver-typescript'),
        ]),
      );
    });
  });

  describe('groupRoutinesByPriority()', () => {
    beforeEach(() => {
      // RunCommandRoutine is mocked, so use plain objects
      routine.routines = [
        { key: 'primary' },
        { key: 'foo' },
        { key: 'bar' },
        { key: 'baz' },
        { key: 'qux' },
      ];
    });

    it('returns all routines if no priority', () => {
      expect(routine.groupRoutinesByPriority()).toEqual({
        other: [{ key: 'primary' }, { key: 'foo' }, { key: 'bar' }, { key: 'baz' }, { key: 'qux' }],
        priority: [],
      });
    });

    it('extracts priority in order defined', () => {
      routine.context.args.priority = 'qux,foo';

      expect(routine.groupRoutinesByPriority()).toEqual({
        other: [{ key: 'primary' }, { key: 'bar' }, { key: 'baz' }],
        priority: [{ key: 'qux' }, { key: 'foo' }],
      });
    });
  });
});
