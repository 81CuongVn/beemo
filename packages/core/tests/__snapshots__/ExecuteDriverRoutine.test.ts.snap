// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteDriverRoutine bootstrap() workspaces errors if workspaces config is empty 1`] = `"Option --workspaces=* provided but project is not workspaces enabled."`;

exports[`ExecuteDriverRoutine bootstrap() workspaces errors if workspaces config is not set 1`] = `"Option --workspaces=* provided but project is not workspaces enabled."`;
